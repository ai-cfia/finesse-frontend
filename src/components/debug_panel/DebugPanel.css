.debug-panel {
  background-color: #f9f9f9;
  border: 1px solid #e0e0e0;
  padding: 15px;
  border-radius: 5px;
  min-width: 250px; /* Adjusted width */
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
  position: absolute;
  top: 150px;
  left: 0px;
  z-index: 1000;
  white-space: normal; /* Updated white-space property */
}

.radio-container {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  margin-bottom: 10px;
}

.radio-option {
  display: flex;
  align-items: center;
  margin-bottom: 10px; 
}

.radio-option input[type="radio"] {
  margin-right: 5px;
  cursor: pointer;
}

.radio-label {
  white-space: nowrap;
  cursor: pointer;
}
