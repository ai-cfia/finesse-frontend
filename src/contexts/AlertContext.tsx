import React, {
  createContext,
  useContext,
  useState,
  type ReactNode,
} from "react";

// Type for the alert context state
interface AlertContextState {
  alertMessage: string | null;
  setAlertMessage: React.Dispatch<React.SetStateAction<string | null>>;
}

// Create the context with a default value
const AlertContext = createContext<AlertContextState>({
  alertMessage: null,
  setAlertMessage: () => {},
});

// Create a provider component for the alert context
export const AlertProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [alertMessage, setAlertMessage] = useState<string | null>(null);

  return (
    <AlertContext.Provider value={{ alertMessage, setAlertMessage }}>
      {children}
    </AlertContext.Provider>
  );
};

// Custom hook to use the alert context
export const useAlert = (): AlertContextState => {
  const context = useContext(AlertContext);
  if (context === null) {
    throw new Error("useAlert must be used within an AlertProvider");
  }
  return context;
};
