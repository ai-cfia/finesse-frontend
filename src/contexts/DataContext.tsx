import { createContext, useContext, useState, type ReactNode } from "react";
import { SearchSource, type QueryResult } from "../types";

// Define the type for your context state
interface DataContextState {
  searchTerm: string | null;
  setSearchTerm: React.Dispatch<React.SetStateAction<string | null>>;
  currentSearchSource: SearchSource;
  setCurrentSearchSource: React.Dispatch<React.SetStateAction<SearchSource>>;
  queryResult: QueryResult[];
  setQueryResult: React.Dispatch<React.SetStateAction<QueryResult[]>>;
}

// Create the context with a default value
const DataContext = createContext<DataContextState>({
  searchTerm: null,
  setSearchTerm: () => {},
  currentSearchSource: SearchSource.Simulated,
  setCurrentSearchSource: () => {},
  queryResult: [],
  setQueryResult: () => {},
});

// Create a provider component
export const DataProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [searchTerm, setSearchTerm] = useState<string | null>(null);
  const [currentSearchSource, setCurrentSearchSource] = useState<SearchSource>(
    SearchSource.Simulated,
  );
  const [queryResult, setQueryResult] = useState<QueryResult[]>([]);

  return (
    <DataContext.Provider
      value={{
        searchTerm,
        setSearchTerm,
        currentSearchSource,
        setCurrentSearchSource,
        queryResult,
        setQueryResult,
      }}
    >
      {children}
    </DataContext.Provider>
  );
};

// Custom hook to use the layout context
export const useData = (): DataContextState => {
  const context = useContext(DataContext);
  if (context === null) {
    throw new Error("useDataFetch must be used within a DataFetchProvider");
  }
  return context;
};
