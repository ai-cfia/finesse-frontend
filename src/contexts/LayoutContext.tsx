import React, {
  createContext,
  useContext,
  useEffect,
  useState,
  type ReactNode,
} from "react";

// Type for your context state
interface LayoutContextState {
  isDebugPanelVisible: boolean;
  toggleDebugPanel: () => void;
  isDebugButtonVisible: boolean;
  isAlertBannerVisible: boolean;
  setIsAlertBannerVisible: React.Dispatch<React.SetStateAction<boolean>>;
}

// Create the context with a default value
const LayoutContext = createContext<LayoutContextState>({
  isDebugPanelVisible: false,
  toggleDebugPanel: () => {},
  isDebugButtonVisible: false,
  isAlertBannerVisible: false,
  setIsAlertBannerVisible: () => {},
});

// Create a provider component
export const LayoutProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [isDebugPanelVisible, setIsDebugPanelVisible] = useState(false);
  const [isDebugButtonVisible, setIsDebugButtonVisible] = useState(false);
  const [isAlertBannerVisible, setIsAlertBannerVisible] = useState(false);

  useEffect(() => {
    const debugMode = process.env.REACT_APP_DEBUG_MODE;
    setIsDebugButtonVisible(debugMode?.toLowerCase() === "true");
  }, []);

  const toggleDebugPanel = (): void => {
    console.log("Toggling Debug Panel visibility");
    setIsDebugPanelVisible(!isDebugPanelVisible);
  };

  return (
    <LayoutContext.Provider
      value={{
        isDebugPanelVisible,
        toggleDebugPanel,
        isDebugButtonVisible,
        isAlertBannerVisible,
        setIsAlertBannerVisible,
      }}
    >
      {children}
    </LayoutContext.Provider>
  );
};

// Custom hook to use the layout context
export const useLayout = (): LayoutContextState => {
  const context = useContext(LayoutContext);
  if (context === null) {
    throw new Error("useLayout must be used within a LayoutProvider");
  }
  return context;
};
