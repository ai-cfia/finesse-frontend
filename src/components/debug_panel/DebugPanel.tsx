import * as React from "react";
import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { fetchFilenames } from "../../api/useApiUtil";
import { useData } from "../../contexts/DataContext";
import { useLayout } from "../../contexts/LayoutContext";
import { SearchSource } from "../../types";
import "./DebugPanel.css";

interface DebugPanelProps {
  onSetSimulatedData?: (data: any) => void;
}

export const DebugPanel: React.FC<DebugPanelProps> = () => {
  const { isDebugPanelVisible } = useLayout();
  const { currentSearchSource, setCurrentSearchSource, setSearchTerm } =
    useData();
  const [filenames, setFilenames] = useState<string[]>([]);
  const navigate = useNavigate();

  useEffect(() => {
    if (currentSearchSource === SearchSource.Simulated) {
      const loadFilenames = async (): Promise<void> => {
        const fetchedFilenames = await fetchFilenames();
        setFilenames(fetchedFilenames);
      };

      loadFilenames()
        .then(() => {
          console.log("Filenames loaded successfully");
        })
        .catch((error) => {
          console.error("Error loading filenames", error);
        });
    } else {
      setFilenames([]);
    }
  }, [currentSearchSource]);

  const handleFilenameClick = (
    e: React.MouseEvent<HTMLButtonElement>,
    filename: string,
  ): void => {
    e.preventDefault();
    // TODO: revisit navigate.
    navigate("/search");
    setSearchTerm(filename);
    alert("Search query set successfully!");
  };

  if (!isDebugPanelVisible) return null;
  return (
    <div className="debug-panel" data-testid="debug-panel">
      <h4>Debug Panel</h4>
      <div className="radio-container">
        <label className="radio-option">
          <input
            type="radio"
            value="ailab"
            checked={currentSearchSource === SearchSource.Ailab}
            onChange={() => {
              setCurrentSearchSource(SearchSource.Ailab);
            }}
          />
          <span className="radio-label">Use AI Lab search</span>
        </label>
        <label className="radio-option">
          <input
            type="radio"
            value="azure"
            checked={currentSearchSource === SearchSource.Azure}
            onChange={() => {
              setCurrentSearchSource(SearchSource.Azure);
            }}
          />
          <span className="radio-label">Use Azure AI Search</span>
        </label>
        <label className="radio-option">
          <input
            type="radio"
            value="simulated"
            checked={currentSearchSource === SearchSource.Simulated}
            onChange={() => {
              setCurrentSearchSource(SearchSource.Simulated);
            }}
          />
          <span className="radio-label">Use Simulated Data</span>
        </label>
      </div>
      {currentSearchSource === SearchSource.Simulated && (
        <div className="input-container">
          <div>
            <h5>Filenames:</h5>
            <ul>
              {filenames.map((filename, index) => (
                <li key={index}>
                  <button
                    onClick={(e) => {
                      handleFilenameClick(e, filename);
                    }}
                  >
                    {filename}
                  </button>
                </li>
              ))}
            </ul>
          </div>
        </div>
      )}
    </div>
  );
};
