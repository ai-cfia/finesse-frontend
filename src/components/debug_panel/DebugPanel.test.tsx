import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { DataProvider } from "../../contexts/DataContext";
import { LayoutProvider } from "../../contexts/LayoutContext";
import { DebugPanel } from "./DebugPanel";

jest.mock("../../api/useApiUtil", () => ({
  fetchFilenames: async () =>
    await Promise.resolve(["file1", "file2", "file3"]),
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
}));

jest.mock("../../contexts/LayoutContext", () => ({
  useLayout: () => ({ isDebugPanelVisible: true }),
  LayoutProvider: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
}));

describe("DebugPanel Component Tests", () => {
  const renderDebugPanel = (): HTMLElement => {
    render(
      <BrowserRouter>
        <DataProvider>
          <LayoutProvider>
            <DebugPanel />
          </LayoutProvider>
        </DataProvider>
      </BrowserRouter>,
    );
    return screen.getByTestId("debug-panel");
  };

  beforeAll(() => {
    window.alert = jest.fn();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test("renders radio buttons for search sources", async () => {
    const view = renderDebugPanel();
    await waitFor(() => {
      const radioButtons = screen.queryAllByRole("radio");
      expect(radioButtons.length).toBe(3);
    });
    expect(view).toMatchSnapshot();
  });

  test("can switch to simulated data and display filenames", async () => {
    const view = renderDebugPanel();

    expect(screen.queryByText("file1")).not.toBeInTheDocument();
    expect(screen.queryByText("file2")).not.toBeInTheDocument();
    expect(screen.queryByText("file3")).not.toBeInTheDocument();

    const simulatedRadioButton = screen.getByLabelText("Use Simulated Data");
    fireEvent.click(simulatedRadioButton);

    await waitFor(() => {
      expect(screen.getByText("file1")).toBeInTheDocument();
    });
    expect(view).toMatchSnapshot();
  });

  test("clicking on a filename button updates the search term", async () => {
    const view = renderDebugPanel();

    expect(screen.queryByText("file1")).not.toBeInTheDocument();

    const simulatedRadioButton = screen.getByLabelText("Use Simulated Data");
    fireEvent.click(simulatedRadioButton);

    const firstFilenameButton = await screen.findByText("file1");
    fireEvent.click(firstFilenameButton);
    expect(window.alert).toHaveBeenCalledWith("Search query set successfully!");

    expect(view).toMatchSnapshot();
  });
});
