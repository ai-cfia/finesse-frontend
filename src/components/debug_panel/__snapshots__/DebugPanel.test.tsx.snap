// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DebugPanel Component Tests can switch to simulated data and display filenames 1`] = `
<div
  class="debug-panel"
  data-testid="debug-panel"
>
  <h4>
    Debug Panel
  </h4>
  <div
    class="radio-container"
  >
    <label
      class="radio-option"
    >
      <input
        checked=""
        type="radio"
        value="ailab"
      />
      <span
        class="radio-label"
      >
        Use AI Lab search
      </span>
    </label>
    <label
      class="radio-option"
    >
      <input
        type="radio"
        value="azure"
      />
      <span
        class="radio-label"
      >
        Use Azure AI Search
      </span>
    </label>
    <label
      class="radio-option"
    >
      <input
        type="radio"
        value="simulated"
      />
      <span
        class="radio-label"
      >
        Use Simulated Data
      </span>
    </label>
  </div>
  <div
    class="input-container"
  >
    <div>
      <h5>
        Filenames:
      </h5>
      <ul>
        <li>
          <button>
            file1
          </button>
        </li>
        <li>
          <button>
            file2
          </button>
        </li>
        <li>
          <button>
            file3
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`DebugPanel Component Tests clicking on a filename button updates the search term 1`] = `
<div
  class="debug-panel"
  data-testid="debug-panel"
>
  <h4>
    Debug Panel
  </h4>
  <div
    class="radio-container"
  >
    <label
      class="radio-option"
    >
      <input
        checked=""
        type="radio"
        value="ailab"
      />
      <span
        class="radio-label"
      >
        Use AI Lab search
      </span>
    </label>
    <label
      class="radio-option"
    >
      <input
        type="radio"
        value="azure"
      />
      <span
        class="radio-label"
      >
        Use Azure AI Search
      </span>
    </label>
    <label
      class="radio-option"
    >
      <input
        type="radio"
        value="simulated"
      />
      <span
        class="radio-label"
      >
        Use Simulated Data
      </span>
    </label>
  </div>
  <div
    class="input-container"
  >
    <div>
      <h5>
        Filenames:
      </h5>
      <ul>
        <li>
          <button>
            file1
          </button>
        </li>
        <li>
          <button>
            file2
          </button>
        </li>
        <li>
          <button>
            file3
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`DebugPanel Component Tests renders radio buttons for search sources 1`] = `
<div
  class="debug-panel"
  data-testid="debug-panel"
>
  <h4>
    Debug Panel
  </h4>
  <div
    class="radio-container"
  >
    <label
      class="radio-option"
    >
      <input
        checked=""
        type="radio"
        value="ailab"
      />
      <span
        class="radio-label"
      >
        Use AI Lab search
      </span>
    </label>
    <label
      class="radio-option"
    >
      <input
        type="radio"
        value="azure"
      />
      <span
        class="radio-label"
      >
        Use Azure AI Search
      </span>
    </label>
    <label
      class="radio-option"
    >
      <input
        type="radio"
        value="simulated"
      />
      <span
        class="radio-label"
      >
        Use Simulated Data
      </span>
    </label>
  </div>
</div>
`;
